Make ChainApi.processHeaders() return a failed future in the case we …

Christewart opened the pull request:
…have no valid headers

fixes #2365

Previously we would _always_ return a successful future from `ChainApi.processHeaders()`. In places like [`DataMessageHandler`](https://github.com/bitcoin-s/bitcoin-s/blob/7172b4a049b8a18568e5e69d3f947ea6b32d4803/node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala#L244) we were assuming that if we received all invalid headers, we would have a failed future returned. Since this was incorrect, we would execute callbacks registered even in the headers were totally invalid.

In this particular case, the wallet project would see that we received a new header, and then would call out to the chain project to retrieve information about that header. Since the chain project  realized that particular header was invalid, it wouldn't store it. However it would return sucessful future like it had.

Christewart committed:
Make ChainApi.processHeaders() return a failed future in the case we have no valid headers

benthecarman commented:
If we get no valid headers, will it cause the node to stop or can it still continue syncing?

Christewart committed:
Run scalafmt

benthecarman commented:
CI failing

Christewart commented:
> CI failing

So the test case that is causing failures here: 

https://github.com/bitcoin-s/bitcoin-s/blob/19c68b77b9736ac403e0730e949d1c7fa57468a0/chain-test/src/test/scala/org/bitcoins/chain/blockchain/ChainHandlerTest.scala#L170

and this is being caused by this method I believe

https://github.com/bitcoin-s/bitcoin-s/blob/40f46d7c7a82a300bf2be2f1e3b957789a80df2d/chain/src/main/scala/org/bitcoins/chain/models/BlockHeaderDAO.scala#L304

In the case of a contentious tip, this does _not_ select all possible tips. This _only_ selects tips with the most work if I understand correctly

benthecarman reviewed (commented):
LGTM barring we get a fix for #2437 first

benthecarman commented on chain-test/src/test/scala/org/bitcoins/chain/blockchain/ChainHandlerTest.scala:
```diff
@@ -94,6 +94,17 @@ class ChainHandlerTest extends ChainDbUnitTest {
+          result <- chainHandler.processHeader(newHeader)
+        } yield result
+      }
+
```
nit: spacing

benthecarman commented:
This should fix #2263 as well

benthecarman commented:
Tried syncing on testnet, I was syncing headers and got to `1484000`. A block was mined on testnet and it was broadcast to me. That caused my entire node to stop syncing because being unable to process it stopped it.

This also means someone could send us a header with bad pow to cause the node to stop syncing.

Edit: made #2460 this should fix the current test failures and the issue I described above

Christewart committed:
Fix test case to check if promise is completed yet

Christewart committed:
WIP: Get something working that isn't network specific

Start putting things back in place

Add comment

Revert logback file

Remove BitcoinSLogger object

Christewart committed:
Rename NodeUnitTest.confg -> NodeUnitTest.getFreshConfig()

Christewart committed:
Rename CachedBitcoinSAppConfig.config -> CachedBitcoinSAppConfig.cachedConfig

Christewart committed:
Make CachedChainAppConfig extend CachedBitcoinSAppConfig

Christewart committed:
Merge branch 'master' of github.com:bitcoin-s/bitcoin-s

Christewart committed:
Fix unused import

Christewart committed:
Get rid of annoying diff

Christewart committed:
Fix spacing nit
